### PR TITLE
Added logging per Brain of time to update policy, time elapsed during training, time to collect experiences, buffer length, average return per policy

### PR DESCRIPTION
**Pull request by eshvk**
Squashing and rebasing against develop

### CODE REVIEW

**Comment by harperj** [comment]
Looks like you have a couple of test failures.

File: ml-agents/mlagents/trainers/trainer.py [hunk 1]
```diff
@@ -17,6 +19,88 @@ class UnityTrainerException(UnityException):
     pass
 
 
+class TrainerMetrics(object):
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer.py]: I'd suggest we move this class into a separate file and add some unit tests for its behavior.

Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: I think the class is fine where it is. It is a helper function that is tied to `Trainer`

File: ml-agents/mlagents/trainers/trainer.py [hunk 2]
```diff
@@ -98,6 +194,7 @@ def get_action(self, curr_info: BrainInfo) -> ActionInfo:
         :param curr_info: Current BrainInfo.
         :return: The ActionInfo given by the policy given the BrainInfo.
         """
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer.py]: It is not obvious to me that this is where the timer should start.

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 3]
```diff
@@ -337,7 +338,7 @@ def update_policy(self):
             self.stats['Losses/Forward Loss'].append(np.mean(forward_total))
             self.stats['Losses/Inverse Loss'].append(np.mean(inverse_total))
         self.training_buffer.reset_update_buffer()
-
+        self.trainer_metrics.end_policy_update()
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/ppo/trainer.py]: I feel like a timer should start here, or else there might be some time lost...

File: ml-agents/mlagents/trainers/trainer.py (hunk 1 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/trainer.py]: I would like it in a separate file as well...

File: ml-agents/mlagents/trainers/trainer.py [hunk 4]
```diff
@@ -17,6 +19,88 @@ class UnityTrainerException(UnityException):
...
+    """
+
+    def __init__(self, path: str, brain_name: str):
+        """
+        :str path: Fully qualified path where CSV is stored.
+        :str brain_name: Identifier for the Brain which we are training
+        """
+        self.path = path
+        self.brain_name = brain_name
+        self.FIELD_NAMES = ['Brain name', 'Time to update policy',
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer.py]: Why not make it static ? Put it on line 27 and it will be a static variable common to all of the implementations of this class.

File: ml-agents/mlagents/trainers/trainer.py [hunk 5]
```diff
@@ -17,6 +19,88 @@ class UnityTrainerException(UnityException):
...
+        self.last_mean_return = mean_return
+        self.time_policy_update_start = time()
+
+    def end_policy_update(self):
+        """
+        Inform Metrics class that policy update has started.
+        """
+        self.delta_policy_update = time() - self.time_policy_update_start
+        delta_train_start = time() - self.time_training_start
+        logger.debug(" Policy Update Training Metrics for {}: "
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer.py]: I would put the second part of this method (starting at this line) into a helper method.

File: ml-agents/mlagents/trainers/trainer_controller.py [hunk 6]
```diff
@@ -44,6 +45,7 @@ def __init__(self, model_path: str, summaries_dir: str,
 
         self.model_path = model_path
         self.summaries_dir = summaries_dir
+        self.training_stats = []
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer_controller.py]: I cannot find where this variable is used.

File: ml-agents/mlagents/trainers/bc/offline_trainer.py [hunk 7]
```diff
@@ -15,7 +15,8 @@
 class OfflineBCTrainer(BCTrainer):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/bc/offline_trainer.py]: Does the new CSV thing works with BC as well or just PPO ?

File: ml-agents/mlagents/trainers/trainer_controller.py [hunk 8]
```diff
@@ -286,14 +298,15 @@ def take_step(self, env, curr_info: AllBrainInfo):
                 # Perform gradient descent with experience buffer
                 trainer.update_policy()
             # Write training statistics to Tensorboard.
+            delta_train_start = time() - self.training_start_time
             if self.meta_curriculum is not None:
                 trainer.write_summary(
                     self.global_step,
-                    lesson_num=self.meta_curriculum
+                    delta_train_start,lesson_num=self.meta_curriculum
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer_controller.py]: missing a space after the coma.

File: ml-agents/tests/trainers/test_learn.py [hunk 9]
```diff
@@ -20,6 +20,7 @@ def basic_options():
         '--slow': False,
         '--no-graphics': False,
         '<trainer-config-path>': 'basic_path',
+        # '--debug': False,
```
Comment (vincentpierre) [inline ml-agents/tests/trainers/test_learn.py]: Dead code ?

File: ml-agents/mlagents/trainers/learn.py [hunk 10]
```diff
@@ -79,7 +78,8 @@ def run_training(sub_id: int, run_seed: int, run_options, process_queue):
     tc = TrainerController(model_path, summaries_dir, run_id + '-' + str(sub_id),
                            save_freq, maybe_meta_curriculum,
                            load_model, train_model,
-                           keep_checkpoints, lesson, external_brains, run_seed)
+                           keep_checkpoints, lesson, env.external_brains,
```
[the remark this answers is not in the document]
Comment (eshvk) [reply ml-agents/mlagents/trainers/learn.py]: Yeah, that will be fixed

File: ml-agents/mlagents/trainers/trainer.py (hunk 5 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: will be fixed

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 11]
```diff
@@ -275,6 +271,8 @@ def process_experiences(self, current_info: AllBrainInfo, new_info: AllBrainInfo
 
                 self.training_buffer[agent_id].reset_agent()
                 if info.local_done[l]:
+                    self.cumulative_returns_since_policy_update.append(self.
+                                                                cumulative_rewards.get(agent_id, 0))
```
[the remark this answers is not in the document]
Comment (eshvk) [reply ml-agents/mlagents/trainers/ppo/trainer.py]: talked offline

File: ml-agents/mlagents/trainers/trainer.py (hunk 1 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: will be fixed

File: ml-agents/mlagents/trainers/bc/offline_trainer.py (hunk 7 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/bc/offline_trainer.py]: Just PPO

File: docs/Training-ML-Agents.md [hunk 12]
```diff
@@ -146,6 +146,11 @@ environment, you can set the following command line options when invoking
   training doesn't involve visual observations (reading from Pixels). See
   [here](https://docs.unity3d.com/Manual/CommandLineArguments.html) for more
   details.
+* `--debug` - Specify this option to run ML-Agents in debug mode and log Trainer
+  Metrics to a CSV stored in the `summaries` directory. The metrics  stored are:
+  brain name, Time to update policy, Time since start of training, Time for last experience collection, Number of experiences used for training, Mean return. This
```
Comment (vincentpierre) [inline docs/Training-ML-Agents.md]: Weird capitalization on this line. If you want to keep the capitalization I would wrap each metric name like this \`name\`

File: docs/Training-ML-Agents.md [hunk 13]
```diff
@@ -146,6 +146,11 @@ environment, you can set the following command line options when invoking
...
+  option is not available currently for Imitation Learning.
+  `
```
Comment (vincentpierre) [inline docs/Training-ML-Agents.md]: Why is there a \` here ?

File: ml-agents/mlagents/trainers/bc/trainer.py [hunk 14]
```diff
@@ -28,22 +28,20 @@ def __init__(self, brain, trainer_parameters, training, load, seed, run_id):
...
 
-        self.summary_path = trainer_parameters['summary_path']
         self.batches_per_epoch = trainer_parameters['batches_per_epoch']
-        if not os.path.exists(self.summary_path):
-            os.makedirs(self.summary_path)
+
 
         self.demonstration_buffer = Buffer()
         self.evaluation_buffer = Buffer()
-        self.summary_writer = tf.summary.FileWriter(self.summary_path)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/bc/trainer.py]: Did you test that BC is still training and logging tensorboard ?

File: ml-agents/mlagents/trainers/trainer.py [hunk 15]
```diff
@@ -98,7 +109,10 @@ def get_action(self, curr_info: BrainInfo) -> ActionInfo:
         :param curr_info: Current BrainInfo.
         :return: The ActionInfo given by the policy given the BrainInfo.
         """
-        return self.policy.get_action(curr_info)
+        self.trainer_metrics.start_experience_collection_timer()
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer.py]: In BC trainer that inherits from trainer but no trainer_metrics is used, will this cause problems ?

File: ml-agents/mlagents/trainers/trainer_controller.py [hunk 16]
```diff
@@ -270,30 +283,35 @@ def take_step(self, env, curr_info: AllBrainInfo):
...
             vector_action=take_action_vector,
             memory=take_action_memories,
             text_action=take_action_text,
             value=take_action_value
         )
-
+        delta_time_step = time() - time_start_step
         for brain_name, trainer in self.trainers.items():
+            if brain_name in self.trainer_metrics:
+                self.trainer_metrics[brain_name].add_delta_step(delta_time_step)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer_controller.py]: Why not `self.trainers[brain_name].trainer_metrics` ?

File: ml-agents/mlagents/trainers/trainer_controller.py [hunk 17]
```diff
@@ -152,7 +162,10 @@ def initialize_trainers(self, trainer_config):
                         .brains_to_curriculums[brain_name]
                         .min_lesson_length if self.meta_curriculum else 0,
                     trainer_parameters_dict[brain_name],
-                    self.train_model, self.load_model, self.seed, self.run_id)
+                    self.train_model, self.load_model, self.seed,
+                    self.run_id)
+                if hasattr(self.trainers[brain_name], 'trainer_metrics'):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer_controller.py]: Don't they all have the attribute sine trainer_metrics is declared in the abstract trainer?

File: ml-agents/mlagents/trainers/trainer_controller.py [hunk 18]
```diff
@@ -270,30 +283,35 @@ def take_step(self, env, curr_info: AllBrainInfo):
...
             if self.meta_curriculum is not None:
                 trainer.write_summary(
                     self.global_step,
-                    lesson_num=self.meta_curriculum
+                    delta_train_start, lesson_num=self.meta_curriculum
                         .brains_to_curriculums[brain_name]
                         .lesson_num)
             else:
-                trainer.write_summary(self.global_step)
+                trainer.write_summary(self.global_step, delta_train_start)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer_controller.py]: What about passing delta_time_step to this guy ? Instead of creating `add_delta_step`

File: ml-agents/mlagents/trainers/trainer_metrics.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+FIELD_NAMES = ['Brain name', 'Time to update policy',
+               'Time since start of training', 'Time for last experience collection',
+               'Number of experiences used for training', 'Mean return']
+
+class TrainerMetrics:
+    """
+        Helper class to track, write training metrics. Tracks time since object
+        of this class is initialized.
+    """
+
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer_metrics.py]: Extra white space ?

File: ml-agents/mlagents/trainers/trainer_metrics.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+    def _add_row(self, delta_train_start):
+        row = [self.brain_name]
+        row.extend(format(c, '.3f') if isinstance(c, float) else c
+                   for c in [self.delta_policy_update, delta_train_start,
+                             self.delta_last_experience_collection,
+                             self.last_buffer_length, self.last_mean_return])
+        self.delta_last_experience_collection = None
+        self.rows.append(row)
+
+
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer_metrics.py]: Extra line

File: docs/Training-ML-Agents.md (hunk 12 above)
Comment (eshvk) [reply docs/Training-ML-Agents.md]: They were taken from the header of the CSV, doesn't matter, I will "normalize" them.

File: ml-agents/mlagents/trainers/trainer.py (hunk 15 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: No because `trainer_metrics` is of type `TrainerMetrics` which is instantiated in the constructor for `Trainer`, the constructor is called in `BCTrainer` so the object will exist and the method can be called. We will not write the data because it will be gibberish.

File: ml-agents/mlagents/trainers/trainer_controller.py (hunk 17 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer_controller.py]: Yep, this was an artifact of our older design where the `trainer_metric` was part of just `PPOTrainer`.  I will fix.

File: ml-agents/mlagents/trainers/trainer_controller.py (hunk 18 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer_controller.py]: Overloads responsibility? I can see a situation where we can't avoid having this in `TrainerController`. I want to eliminate `add_delta_step` at some point post a cleaner refactor?

File: ml-agents/mlagents/trainers/trainer_controller.py (hunk 16 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer_controller.py]: Can you explain this a bit more? Are you saying why have the if condition?

File: ml-agents/mlagents/trainers/bc/trainer.py (hunk 14 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/bc/trainer.py]: Just tried it, looks like it is still running and logging

**Comment by eshvk** [comment]
PTAL @vincentpierre Fixed most of the outstanding issues, squashed, rebased from develop
